aws_eks: New module for managing AWS EKS

willthames opened the pull request:
##### SUMMARY
aws_eks module is used for creating and removing EKS clusters.

Includes full test suite and updates to IAM policies to enable it.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aws_eks

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (devel a60fe1946c) last updated 2018/06/06 22:11:31 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Mar  9 2018, 23:57:12) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
There seems to be a bug with ec2_group such that the test suite thinks it has removed the security groups but they still exist which means that the VPC tear down doesn't work correctly. However, that is not a bug in this module.

There are no waiters for this module at present in boto - we might need to add one if we choose to provide `wait` functionality, but that can be a future improvement.

`aws_eks_facts` shouldn't be too far behind this PR

s-hertel commented on test/integration/targets/aws_eks/playbooks/roles/aws_eks/tasks/main.yml:
```diff
@@ -0,0 +1,193 @@
+
+    - name: create subnets
+      ec2_vpc_subnet:
+        az: '{{ ec2_region }}{{ item.zone }}'
```
s/ec2_region/aws_region

s-hertel commented on test/integration/targets/aws_eks/playbooks/roles/aws_eks/tasks/main.yml:
```diff
@@ -0,0 +1,193 @@
+      with_items: "{{ eks_subnets }}"
+      ignore_errors: yes
+
+    - name: remove setup VPC
```
Can you add some retries to this? I've run it a few times and gotten 
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_gfbyo1jp/ansible_module_ec2_vpc_net.py", line 467, in main
    connection.delete_vpc(VpcId=vpc_id)
  File "/tmp/ansible_gfbyo1jp/ansible_modlib.zip/ansible/module_utils/aws/core.py", line 241, in deciding_wrapper
    return unwrapped(*args, **kwargs)
  File "/Users/shertel/Workspace/ansible/venv/python3.6.0/lib/python3.6/site-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/shertel/Workspace/ansible/venv/python3.6.0/lib/python3.6/site-packages/botocore/client.py", line 612, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (DependencyViolation) when calling the DeleteVpc operation: The vpc 'vpc-18020e61' has dependencies and cannot be deleted.
```
and had to wait and then delete the VPC myself.

s-hertel replied:
This is an ec2_group bug.

s-hertel reviewed (commented):
Looks great!

Commit:
aws_eks: New module for managing AWS EKS

aws_eks module is used for creating and removing EKS clusters.

Includes full test suite and updates to IAM policies to enable it.

s-hertel commented on test/integration/targets/aws_eks/playbooks/roles/aws_eks/defaults/main.yml:
```diff
@@ -0,0 +1,33 @@
+  - name: "{{ eks_cluster_name }}-control-plane-sg"
+    description: "EKS Control Plane Security Group"
+    rules:
+      - group_name: "{{ eks_cluster_name }}-workers-sg"
```
Sorry, back and forth with myself over this. I think deleting the VPC is failing because this security group never gets cleaned up. "{{ eks_cluster_name }}-worker-sg" and "{{ eks_cluster_name }}-control-plane-sg" get deleted but not this one.

s-hertel committed:
Clean up all security groups

s-hertel committed:
appease shippable

ryansb reviewed (approved):
Code+interface is great, I think it would be good to rename to `aws_eks_cluster` since I think EKS will add features+resources, requiring additional EKS modules.

willthames committed:
Rename aws_eks module to aws_eks_cluster

deiwin commented:
@willthames, do you know if there's been any progress on the `wait` functionality?

willthames commented:
@deiwin I haven't made any progress - I've been working on other things.

deiwin commented:
Ok, thanks. I need that functionality, so I'll likely implement it based on [`waiters.py` here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/waiters.py). Let me know if you think there's a better way.
